### ISSUE TITLE
React Hooks (Hooks can only be called inside the body of a function component)

### ISSUE THREAD

**Issue by lukaskamp** [shakacode/react_on_rails#1198]
I'm trying Testing React Hooks with react_on_rails gem and noticed there is a Error: 

"Hooks can only be called inside the body of a function component error."

Without using ReactOnRails config (e.g in packs/applications.js) everything works correctly. Also when I used react-rails gem everything works. 

Where could be problem ?

**Comment by amauryfischer** [comment]
Got same problem there, i tried a couple of react suggestion(veryfing i'm using hooks correctly, and veryfing that i got a single version of React) but it's not one of them...
Seems we are not alone, don't know if it comes from rails webpacker, React on rails, or anything else.
Related issue on rails webpacker : https://github.com/rails/webpacker/issues/1840

**Comment by justin808** [comment]
@lukaskamp @amauryfischer can one of you give me a simple reproduction case? a github repo?

**Comment by wuz** [comment]
As a note, a quick fix for this is to just wrap your React On Rails rendered component in a function, like this:

```
export default props => <ComponentHere {...props} />;
```

**Comment by amauryfischer** [comment]
```
import React,{useState} from 'react'
const Test = function() {
    const [state, setstate] = useState(0);
    return (
        <div onClick={setstate(state + 1,)}>{state}</div>
    )
}

export default props => <Test {...props} />;
```

@wuz didn't work, same error :
```
Invariant Violation: Hooks can only be called inside the body of a function component.
```

**Comment by wuz** [comment]
@amauryfischer Is that the exact code you are using? That `setstate` call in the button won't work, since that is just getting called on render, causing an infinite loop.

Using `export default props => <Test {...props} />;` worked for me.

**Comment by amauryfischer** [comment]
Yes my bad, but even corrected
```
import React,{useState} from 'react'
const Test = function() {
    const [state, setstate] = useState(0);
    return (
        <div onClick={() => setCount(count + 1)}>{state}</div>
    )
}

export default props => <Test {...props} />;
```

got the same error.
Even with only <div></div> in the return.
This is so strange ><

**Comment by juliusdelta** [comment]
We're having the same issue.

We have multiple components rendering on the same page. We believe this is related to the issue described [in the react docs](https://reactjs.org/warnings/invalid-hook-call-warning.html) about loading 2 instances of React, possibly.

The easiest way to reproduce is to make 2 components with hooks and register them then pass those two into a Rails view. If I have enough time later today I'll make a reproduction repo and update this comment.

**Comment by amauryfischer** [comment]
@juliusdelta Exactly ! I love you so much. Finally find out why my hooks wasn't working thanks to you.
i had a javascript_pack_tag with a component registered in application.html.erb and it brokes my whole components for each view due to multiple React for each view.

I'm happy, was searching for this since a whole month

**Comment by janklimo** [comment]
I've put together a minimal reproduction repo that can be found [here](https://github.com/janklimo/react-on-rails-hooks). 

As suggested by @wuz above, the error is thrown based on the way we export the component:

```js
// Works ✅
export default props => <HelloWorld {...props} />;

// Fails 🔴
export default HelloWorld;
```
Here's the error message:

![Screen Shot 2019-04-17 at 1 01 42 AM](https://user-images.githubusercontent.com/7811733/56233050-649a8380-60ac-11e9-8d0d-9f1f24e1ab3e.png)

I've tried using the latest `react` and `react-dom` versions but it didn't help. There is only one `react` component embedded on the `hello_world` page. Turbolinks are not included.

I've been recently dealing with this problem in our app (not using`react_on_rails`) using Turbolinks and it appears that rendering the component in both preview and final render causes some sort of race condition. Not rendering `react` components in preview seems to do the trick but it's interesting to see it's not a Turbolinks issue, so happy to help work on this to resolve the root cause.

**Comment by rubiety** [comment]
@janklimo's workaround above using `export default props => <HelloWorld {...props} />;` does indeed seem to work for the case where there is only one `react_component` on the page, and I am using it successfully in that case.

Unfortunately, when the page contains *multiple* `react_component` invocations (for a component using hooks), this appears to *not* work, as is pointed out by @juliusdelta and @amauryfischer. Even creating an explicit wrapper class-based component and rendering that doesn't work either.

So as it stands right now, it appears to just not be possible to use react hooks while rendering multiple components on the same page. This is a pretty serious problem, and seems to mean the only way to work around this is to rewrite components without using hooks *at all* (at any level of the component tree, when using multiple components o the same page).

**Comment by rubiety** [comment]
After some further investigation on this, and reading [this issue](https://github.com/facebook/react/issues/13991) carefully, I realized my particular issue was the inclusion of *multiple bundles on the same page*. The problem with two or more `javascript_pack_tag`'s on the same page is that if more than one bundle includes a copy of react, then this will cause problems with React hooks.

I was able to work around this problem by consolidating multiple packs into one to ensure that only one React copy was loaded. After that, @janklimo's fix using `export default props => <HelloWorld {...props} />;` *does* indeed work around this problem.

**Comment by janklimo** [comment]
Thank you for the update @rubiety 👍 While evaluating our options I tested numerous gems for using React with Rails and `react-rails` seems to have hooks figured out (my [demo](https://react-rails-hooks.herokuapp.com/) and [repo](https://github.com/janklimo/react-rails-demo)). Not meaning to promote alternative gems, just thought it could help identify the problem.

**Comment by justin808** [comment]
> I was able to work around this problem by consolidating multiple packs into one to ensure that only one React copy was loaded. After that, @janklimo's fix using export default props => <HelloWorld {...props} />; does indeed work around this problem.

@rubiety Having multiple copies of React sent to the browser is also bad for performance. So definitely having on copy of React is correct.

@janklimo I really appreciate your help on this issue. If anybody trying to debug this wants to pair with me, please schedule a time with my [calendar link](https://meetings.hubspot.com/justingordon/30-minute-consultation) or [email me](mailto:justin@shakacode.com) for a Slack invite.

**Comment by tahsin352** [comment]
@justin808 I have further checked the problem. several others also faced this issue with react hooks. ref. https://github.com/facebook/react/issues/13991

To solve 'multiple instances of react' error, I have tried some other ways like below:
```
#/config/webpack/custom.js
const path = require('path')

module.exports = {
  resolve: {
    alias: {
      'react': path.resolve('./node_modules/react'),
      'react-dom': path.resolve('./node_modules/react-dom')
    }
  }
}

# config/webpack/environment.js
const { environment } = require('@rails/webpacker')
const customConfig = require('./custom')

environment.config.merge(customConfig)
```
again, I have also tried with npm link package. but it also didn't solve the error.
another solution i tried, using 'resolved_path' of webpacker, no luck. another way is 'nohoist' or 'externals' property of webpak.

Can i symlinked dependencies use their own copy of react in a react_on_rails gem file ???

**Comment by filipkis** [comment]
I had the same problem which occurred when I tried to render a second component. Few hours of debugging later I realised that the problem was not related to having two components actually, but the fact that a root component (the one that I put in my view) was the one using  `useState` hook. 

As soon as I wrapped it in another component that didn't use the hook all was fine with the world again.

In other words, hooks for some reason can't be in root component that you are rendering with `react_on_rails`.

**Comment by tahsin352** [comment]
@filipkis nice finding, thanks

**Comment by batamire** [comment]
I migrated from webpacker 3.6.0 to 4.0.7 and started seeing this error. Maybe this can help track down the issue?
I have only one component in my view, wrapping does work.

**Comment by beeeku** [comment]
Can anyone point me in a direction How can I render a container with react hooks?

**Comment by theocerutti** [comment]
Any news about this ? Because I can't use hooks anywhere..

**Comment by aflansburg** [comment]
We have been noticing this issue as well, it seems to occur possibly when we use `@material-ui/core` and some of its helpers such as `withStyles()` or possibly `makeStyles()` - the fix we are using for now is to wrap our export in a HOC: 

```
export default props => <FunctionComponent {...props} />;
```

**Comment by aflansburg** [comment]
Was going to add, if you come up on this problem just try this:

```
function YourComponent(props){ // implementation }

export default props => <YourComponent {...props} />
```

As of yet I have no idea why that works. Not sure if this project is being actively maintained anymore.

**Comment by jacksonrayhamilton** [comment]
I think the issue is that React on Rails will call a function component like a normal function [here](https://github.com/shakacode/react_on_rails/blob/9aba6741549c70825b17b3207d99d244b55dc449/node_package/src/createReactElement.js#L39), but the hooks code doesn’t expect for components to be used like that; I think it only expects function components to be passed to `React.createElement`.

In light of this @justin808, what do you think about removing this snippet?

```js
  if (generatorFunction) {
    return component(props, railsContext);
  }
```

**Comment by justin808** [comment]
@jacksonrayhamilton React on Rails allows you to create a function that takes 2 params, the props and the railsContext, and then your function should return a React component.

So I'm not totally clear on what you're suggesting.

Can you create a simple example to demonstrate this?

**Comment by jacksonrayhamilton** [comment]
Sure.  In my pack file I have a function component that uses a hook, and I register this component:

```js
import React, { useState } from 'react'

function HelloWorld() {
  const [whom] = useState('Everyone')
  return <div>Hello {whom}!</div>
}

ReactOnRails.register({ HelloWorld })
```

In my Rails template I render it:

```js
<%= react_component('HelloWorld', prerender: false) %>
```

When I load up the page, I see that when my function component is being registered, `ReactOnRails#register` does not distinguish between a function component and a “generator function.”

<img src="https://shareitdammit.com/eMbxbP1V">

Thus when I advance to the error, we can see that React is unhappy about `useState` being called…

<img src="https://shareitdammit.com/PwnjZj6a">

Because a function component was called outside the React rendering process…

<img src="https://shareitdammit.com/9wirBnox">

And thus `useState` was also called outside the React rendering process.

<img src="https://shareitdammit.com/ks6ALrKu">

So I think the key is that “`ReactOnRails#register` does not distinguish between a function component and a ‘generator function.’”  And perhaps that is not even possible to do reliably.

**Comment by dchersey** [comment]
This is nuts ... if anyone wants a working example of HMR with react-rails that supports hooks, look at this:  https://github.com/Leap-Forward/react-rails-hmr .

**Comment by jacksonrayhamilton** [comment]
Well, I don’t think react_on_rails users need to resort to a whole new gem to fix their problem.  I think in my last comment here I narrowed in on the problem and I expect it will be pretty easy to solve in the react_on_rails source.  Maybe with a breaking API change at worst?  I expect @justin808 has just been busy.  Hooks are relatively new.  I respect his choices in prioritizing his time.

**Comment by justin808** [comment]
Awesome to hear from you all. We're working on some updates to the sample apps.

@jacksonrayhamilton I think you hit the nail on the head! The core issue is that we would _like_ a way that we can infer if a function is for a React Component or just a function.

So the source links:

* https://github.com/shakacode/react_on_rails/blob/master/node_package/src/ReactOnRails.ts#L40
* https://github.com/shakacode/react_on_rails/blob/master/node_package/src/ComponentRegistry.ts#L21
* https://github.com/shakacode/react_on_rails/blob/master/node_package/src/generatorFunction.ts#L11

Either we have to fix the `generatorFunction` to be able to detect the difference, or we need to update the API so that we can specify generator function or a simple functional component.

How about if we use [Function.length](the https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/length)?

If the length is 2, then we assume a generator function that will take two args:
1. Props
2. RailsContext

If the length is 3, we already assume that this is a "renderer" which is used for lazy loading.

If the length is zero or 1, then we can assume that this is simple React Component.

Thus, we can then call the [React.createElement API](https://reactjs.org/docs/react-api.html#createelement)

I took a try at this fix:

https://github.com/shakacode/react_on_rails/pull/1268



@behraaangm @Judahmeek @ashgaliyev can you guys please comment?

@jacksonrayhamilton thank you for your contribution!. Do you want a [one-month free subscription to React on Rails Pro](https://forum.shakacode.com/t/react-on-rails-pro-2020-shave-the-yak/1842)? I'm going to formalize this more soon on the main https://www.shakacode.com website.

**Comment by jacksonrayhamilton** [comment]
Thanks for the offer Justin.  We'll take you up on that if we need further help.

> How about if we use Function.length?

I think that could work.

**Comment by sharjeel288** [comment]
if you want to fix this problem just use react:"^16.8.0"  react-dom: "^16.8.0",
    react-scripts: "2.1.1", in package.json

**Comment by Judahmeek** [comment]
@justin808 #1268 closes this issue, correct?

**Comment by justin808** [comment]
Yes it does!

**Comment by justin808** [comment]
@jacksonrayhamilton @sharjeel288 @dchersey @filipkis @lukaskamp @amauryfischer @wuz @juliusdelta @janklimo @rubiety @tahsin352 @batamire @theocerutti @aflansburg 

I'd love feedback:

React on Rails v12: Should we just error out and force the change to wrap the result JSX in a function with zero params?

or should we make the upgrade easier for 99% of the cases that did not use Hooks. 

https://github.com/shakacode/react_on_rails/pull/1280/files#r419985490

Let me know if I can clarify this choice....

Basically, if there is more than one param, I can either force the return value of a React Function Component (which can use hooks) or I can check if the return value is a React Element. If you used hooks calls in the function that created the React Element (JSX), you will get the obscure hooks error.

So....

1. Harder to upgrade, but avoid confusion when using the Hooks API
2. Easier to upgrade, but still supports the Hooks API properly

**Comment by jacksonrayhamilton** [comment]
> Let me know if I can clarify this choice....

Will you please provide some code examples of the API choices you’re offering?

**Comment by justin808** [comment]
If you take 2 or more params, export a React Function Component:

```js
import React from 'react';
import AppComponent from './AppComponent';
const AppComponentWithRailsContext = (props, railsContext) => (
  // It's better to create a React Function Component so you can
  // use the React Hooks API in this React Function Component
  () => <AppComponent {...{...props, railsContext}}/>
)
export default AppComponentWithRailsContext
```

Currently, this works, to just return a React Element (JSX)
```js
import React from 'react';
import AppComponent from './AppComponent';
const AppComponentWithRailsContext = (props, railsContext) => (
  <AppComponent {...{...props, railsContext}}/>
)
export default AppComponentWithRailsContext
```

I propose that I'll allow the second case for v12, but print a **warning** console message indicating that the JSX should get wrapped in a function to make the return value a React Function Component.

**Comment by jacksonrayhamilton** [comment]
I see.  I’m not using `railsContext` in my code so I’m not really affected by this.  But I can sympathize with people who expect the second case to “just work.”

However, if the programmer does try to use a hook in the second case, stinks to be him, because he’ll be back to Square One in terms of this GitHub issue.  I’d lean towards preventing the possibility of that scenario by being strict about the return value.

I see in that PR you’re planning on fully removing the pattern by v13, so I suppose it doesn’t make much of a difference if it’s done later or now.

**Comment by justin808** [comment]
I merged https://github.com/shakacode/react_on_rails/pull/1280.

It console.errors with a detailed message. I'll just throw for the following major version.

I want to ship v12 soon!

**Comment by justin808** [comment]
@jacksonrayhamilton @sharjeel288 @dchersey @filipkis @lukaskamp @amauryfischer @wuz @juliusdelta @janklimo @rubiety @tahsin352 @batamire @theocerutti @aflansburg Any of you want to help me test the master branch? I think we're close.

I could also create a beta version on rubygems and npm if one of you wanted that.

**Comment by justin808** [comment]
@jacksonrayhamilton @sharjeel288 @dchersey @filipkis @lukaskamp @amauryfischer @wuz @juliusdelta @janklimo @rubiety @tahsin352 @batamire @theocerutti @aflansburg

Gem version: 12.0.0.pre.beta.0 and npm version 12.0.0-beta.0 released! Please try this out!

https://www.npmjs.com/package/react-on-rails
https://rubygems.org/gems/react_on_rails

**Comment by justin808** [comment]
Fixed in v12!

**Comment by justin808** [comment]
@jacksonrayhamilton @sharjeel288 @dchersey @filipkis @lukaskamp @amauryfischer @wuz @juliusdelta @janklimo @rubiety @tahsin352 @batamire @theocerutti @aflansburg

Hey, hey, React on Rails v12 has shipped!

**Comment by jacksonrayhamilton** [comment]
Awesome, thanks @justin808!

**Comment by imgarylai** [comment]
I'm running a new Rails app with ruby 2.7.1 and Rails 6.0.3.4. I followed this [tutorial](https://github.com/shakacode/react_on_rails/blob/master/docs/tutorial.md) and got the same error message as well. 

```
Uncaught Invariant Violation: ReactOnRails encountered an error while rendering component: HelloWorld.
Original message: Hooks can only be called inside the body of a function component. (https://fb.me/react-invalid-hook-call)
    at invariant (http://localhost:3000/packs/js/hello-world-bundle-842597fa5db1436cba9a.js:33684:19)
    at resolveDispatcher (http://localhost:3000/packs/js/hello-world-bundle-842597fa5db1436cba9a.js:35069:32)
    at useState (http://localhost:3000/packs/js/hello-world-bundle-842597fa5db1436cba9a.js:35093:24)
    at HelloWorld (http://localhost:3000/packs/js/hello-world-bundle-842597fa5db1436cba9a.js:121:73)
    at createReactElement (http://localhost:3000/packs/js/hello-world-bundle-842597fa5db1436cba9a.js:33323:12)
    at render (http://localhost:3000/packs/js/hello-world-bundle-842597fa5db1436cba9a.js:33132:75)
    at forEach (http://localhost:3000/packs/js/hello-world-bundle-842597fa5db1436cba9a.js:33048:5)
    at forEachComponent (http://localhost:3000/packs/js/hello-world-bundle-842597fa5db1436cba9a.js:33061:3)
    at reactOnRailsPageLoaded (http://localhost:3000/packs/js/hello-world-bundle-842597fa5db1436cba9a.js:33169:3)
    at HTMLDocument.renderInit (http://localhost:3000/packs/js/hello-world-bundle-842597fa5db1436cba9a.js:33194:5)
```

Gemfile: 

```Gemfile
...
gem 'react_on_rails', '12.0.3'
gem 'webpacker', '~> 5.2'
...
```

```
➜  ruby -v
ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-darwin19]
➜  rails -v
Rails 6.0.3.4
```

**Comment by justin808** [comment]
@imgarylai Can you please share your repo on Github. If you're 100% sure this is an issue, please open a new issue.

**Comment by justin808** [comment]
Might be related to React v17. The tutorial steps will get the latest React. 

https://github.com/facebook/react/blob/master/CHANGELOG.md#1700-october-20-2020

**Comment by justin808** [comment]
@imgarylai can you try to adjust the steps so that you don't use React 17 to isolate if that is the cause?

**Comment by imgarylai** [comment]
@justin808 Thank you! I will try it again and use `React 16`

**Comment by jwcatprint** [comment]
@justin808 I received the same error with the demo when installing on a new app

**Comment by justin808** [comment]
> @justin808 I received the same error with the demo when installing on a new app

@jwcatprint are you using React 17? I'll look into this no later than this coming weekend.

In the meantime can somebody confirm that 

```
yarn add react@16.14.0 react-dom@16.14.0
```

solves the issue.

**Comment by justin808** [comment]
PRs welcome!

**Comment by justin808** [comment]
Moving over to a new issue #1336.

**Comment by justin808** [comment]
This should have been fixed in 12.0.4.

**Comment by samtgarson** [comment]
In case anyone else is stuck on the same issue I am (unrelated to other fixes so far):

 I've been getting this on 12.2... turned out to be because of [this issue](https://github.com/rails/webpacker/issues/2932#issuecomment-860590789) which is caused by adding multiple calls to `javascript_pack_tag`s on one page, leading to multiple instances of React. 

For now, the fix is to only call `javascript_pack_tag` once per page.

**Comment by justin808** [comment]
I'm reopening this issue just to make it clear that **javascript_pack_tag** needs to be only once on the view, including the layout when using rails/webpacker v6.

**Comment by vtamara** [comment]
Referencing https://github.com/shakacode/shakapacker/issues/39 because solving that would help here.

**Comment by justin808** [comment]
Fixed a long time ago.
